blockchain/pool.go

```diff
@@ -1,630 +1,372 @@
-	startTime time.Time
-
-	mtx sync.Mutex
-	// block requests
-	requesters map[int64]*bpRequester
-	height     int64 // the lowest key in requesters.
-	// peers
-	peers         map[p2p.ID]*bpPeer
-	maxPeerHeight int64 // the biggest reported height
-
-	// atomic
-	numPending int32 // number of requests pending assignment or block response
-
-	requestsCh chan<- BlockRequest
-	errorsCh   chan<- peerError
+type blockData struct {
+	block *types.Block
+	peer  *bpPeer
 }
 
-// NewBlockPool returns a new BlockPool with the height equal to start. Block
-// requests and errors will be sent to requestsCh and errorsCh accordingly.
-func NewBlockPool(start int64, requestsCh chan<- BlockRequest, errorsCh chan<- peerError) *BlockPool {
-	bp := &BlockPool{
-		peers: make(map[p2p.ID]*bpPeer),
-
-		requesters: make(map[int64]*bpRequester),
-		height:     start,
-		numPending: 0,
-
-		requestsCh: requestsCh,
-		errorsCh:   errorsCh,
+func (bd *blockData) String() string {
+	if bd == nil {
+		return fmt.Sprintf("blockData nil")
 	}
-	bp.BaseService = *cmn.NewBaseService(nil, "BlockPool", bp)
-	return bp
-}
-
-// OnStart implements cmn.Service by spawning requesters routine and recording
-// pool's start time.
-func (pool *BlockPool) OnStart() error {
-	go pool.makeRequestersRoutine()
-	pool.startTime = time.Now()
-	return nil
-}
-
-// spawns requesters as needed
-func (pool *BlockPool) makeRequestersRoutine() {
-	for {
-		if !pool.IsRunning() {
-			break
-		}
-
-		_, numPending, lenRequesters := pool.GetStatus()
-		if numPending >= maxPendingRequests {
-			// sleep for a bit.
-			time.Sleep(requestIntervalMS * time.Millisecond)
-			// check for timed out peers
-			pool.removeTimedoutPeers()
-		} else if lenRequesters >= maxTotalRequesters {
-			// sleep for a bit.
-			time.Sleep(requestIntervalMS * time.Millisecond)
-			// check for timed out peers
-			pool.removeTimedoutPeers()
-		} else {
-			// request for more blocks.
-			pool.makeNextRequester()
+	if bd.block == nil {
+		if bd.peer == nil {
+			return fmt.Sprintf("block: nil peer: nil")
 		}
+		return fmt.Sprintf("block: nil peer: %v", bd.peer.id)
 	}
+	return fmt.Sprintf("block: %v peer: %v", bd.block.Height, bd.peer.id)
 }
 
-func (pool *BlockPool) removeTimedoutPeers() {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-
-	for _, peer := range pool.peers {
-		if !peer.didTimeout && peer.numPending > 0 {
-			curRate := peer.recvMonitor.Status().CurRate
-			// curRate can be 0 on start
-			if curRate != 0 && curRate < minRecvRate {
-				err := errors.New("peer is not sending us data fast enough")
-				pool.sendError(err, peer.id)
-				pool.Logger.Error("SendTimeout", "peer", peer.id,
-					"reason", err,
-					"curRate", fmt.Sprintf("%d KB/s", curRate/1024),
-					"minRate", fmt.Sprintf("%d KB/s", minRecvRate/1024))
-				peer.didTimeout = true
-			}
-		}
-		if peer.didTimeout {
-			pool.removePeer(peer.id)
-		}
+type blockPool struct {
+	logger log.Logger
+	peers  map[p2p.ID]*bpPeer
+	blocks map[int64]p2p.ID
```

Comment:
Can we have a comment explaining what this variable represent?

Reply:
done